### PR TITLE
envoy: Use upstream filters for L7 LB policy enforcement

### PR DESCRIPTION
**Pull request by jrajahalme**
Update `cilium-envoy` to build with dual (downstream/upstream) filter for Cilium HTTP policy enforcement. Use of the upstream filter depends on the bpf metadata listener filter being present, as well as the downstream Cilium HTTP policy enforcement filter taking care of the ingress policy enforcement.

Upstream filter is only needed with L7 LB, as otherwise the final destination address is already known when the downstream filter executes. Cilium Envoy filters still enforce non-L7 LB traffic in the downstream filter chain, so no changes are needed for generating Cilium Envoy config for regular ingress/egress policy enforcement. Hence the changes here only apply to Envoy configuration created for CEC/CCEC. If L7 LB is indicated, then the new upstream l7 policy filter is injected to take care of L7 LB egress policy enforcement after the destination has been selected. Previously we did this from the downstream filter via a custom callback from the upstream side.

Upstream policy enforcement is bypassed for shadow/mirror streams, as required for Gateway API conformance tests.

Related proxy PR: https://github.com/cilium/proxy/pull/581

```release-note
cilium-envoy now uses upstream filter chains for L7 LB policy enforcement.
```

### CODE REVIEW

File: pkg/ciliumenvoyconfig/cec_resource_parser.go [hunk 1]
```diff
@@ -430,6 +443,50 @@ func (r *cecResourceParser) parseResources(cecNamespace string, cecName string,
...
+			if a != nil {
+				if err := a.UnmarshalTo(opts); err != nil {
+					return envoy.Resources{}, fmt.Errorf("failed to unmarshal HttpProtocolOptions")
+				}
+			}
+
+			// Figure out if upstream supports ALPN so that we can add missing upstream
+			// protocol config
+			supportsALPN := false
+			// ts := cluster.GetTransportSocket()
```
Comment (mhofstetter) [inline pkg/ciliumenvoyconfig/cec_resource_parser.go]: removing the commented out code?

File: pkg/ciliumenvoyconfig/cec_resource_parser.go [hunk 2]
```diff
@@ -381,13 +389,18 @@ func (r *cecResourceParser) parseResources(cecNamespace string, cecName string,
...
+
 	// Allocate TPROXY ports for listeners without address.
 	// Do this only after all other possible error cases.
 	for _, listener := range resources.Listeners {
 		// Figure out if this is an internal listener
 		isInternalListener := listener.GetInternalListener() != nil
 
 		if !isInternalListener {
+			// upstream filters are injected if any non-internal listener is L7 LB
+			injectCiliumUpstreamFilters = injectCiliumUpstreamFilters || isL7LB
```
Comment (mhofstetter) [inline pkg/ciliumenvoyconfig/cec_resource_parser.go]: do we have to add the condition whether any downstream cilium filter has been added to the listener? (e.g. edge case if listener address isn't set by Cilium)

hence only add upstream filter if a downstream filter is present? or doesn't this matter?

File: pkg/ciliumenvoyconfig/cec_resource_parser.go [hunk 3]
```diff
@@ -595,6 +652,81 @@ func qualifyHttpFilters(cecNamespace string, cecName string, hcmConfig *envoy_co
...
+		if j >= 0 {
+			filters = append(filters[:j+1], filters[j:]...)
+			filters[j] = envoy.GetCiliumHttpFilter()
+		} else {
+			filters = append(filters, envoy.GetCiliumHttpFilter())
+		}
+		changed = true
+	}
+
+	// Add CodedFilter if missing
```
Comment (mhofstetter) [inline pkg/ciliumenvoyconfig/cec_resource_parser.go]: nit: typo -> CodecFilter

**Review by mhofstetter — COMMENTED** [review]
nice work - some comments/questions inline!

File: pkg/ciliumenvoyconfig/cec_resource_parser.go [hunk 4]
```diff
@@ -430,6 +443,50 @@ func (r *cecResourceParser) parseResources(cecNamespace string, cecName string,
...
+			opts := &envoy_config_upstream.HttpProtocolOptions{}
+
+			if cluster.TypedExtensionProtocolOptions == nil {
+				cluster.TypedExtensionProtocolOptions = make(map[string]*anypb.Any)
+			}
+
+			a := cluster.TypedExtensionProtocolOptions[httpProtocolOptionsType]
+			if a != nil {
+				if err := a.UnmarshalTo(opts); err != nil {
+					return envoy.Resources{}, fmt.Errorf("failed to unmarshal HttpProtocolOptions")
```
Comment (tommyp1ckles) [inline pkg/ciliumenvoyconfig/cec_resource_parser.go]: Should we include err as part of the returned error?

File: pkg/ciliumenvoyconfig/cec_resource_parser.go [hunk 5]
```diff
@@ -595,6 +652,81 @@ func qualifyHttpFilters(cecNamespace string, cecName string, hcmConfig *envoy_co
...
+				UseDownstreamProtocolConfig: &envoy_config_upstream.HttpProtocolOptions_UseDownstreamHttpConfig{
+					Http2ProtocolOptions: &envoy_config_core.Http2ProtocolOptions{},
+				},
+			}
+		}
+		changed = true
+	}
+
+	if err := opts.Validate(); err != nil {
+		return false, fmt.Errorf("failed to validate HttpProtocolOptions after injecting Cilium upstream filters (%w): %s", err, opts.String())
```
Comment (tommyp1ckles) [inline pkg/ciliumenvoyconfig/cec_resource_parser.go]: It might be more idiomatic/intuitive to do input first then error: `(%s): %w, opts.String(), err`

**Review by tommyp1ckles — COMMENTED** [review]
k8s changes look good, just a few minor comments

**Review by gandro — APPROVED** [review]
Helm LGTM

File: pkg/ciliumenvoyconfig/cec_resource_parser.go (hunk 1 above)
Comment (jrajahalme) [reply pkg/ciliumenvoyconfig/cec_resource_parser.go]: I'll uncomment it instead.

File: pkg/ciliumenvoyconfig/cec_resource_parser.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/ciliumenvoyconfig/cec_resource_parser.go]: Good point, must skip injecting upstream filters if downstream filters were not injected!

File: pkg/ciliumenvoyconfig/cec_resource_parser.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/ciliumenvoyconfig/cec_resource_parser.go]: fixed, thanks!

File: pkg/ciliumenvoyconfig/cec_resource_parser.go (hunk 4 above)
Comment (jrajahalme) [reply pkg/ciliumenvoyconfig/cec_resource_parser.go]: Added :-)

File: pkg/ciliumenvoyconfig/cec_resource_parser.go (hunk 5 above)
Comment (jrajahalme) [reply pkg/ciliumenvoyconfig/cec_resource_parser.go]: Right, fixed!

File: pkg/ciliumenvoyconfig/cec_resource_parser.go (hunk 1 above)
Comment (mhofstetter) [reply pkg/ciliumenvoyconfig/cec_resource_parser.go]: even better :)

File: pkg/ciliumenvoyconfig/cec_resource_parser.go (hunk 5 above)
Comment (mhofstetter) [reply pkg/ciliumenvoyconfig/cec_resource_parser.go]: looks like we have at least two more occurrences in the parser that aren't idiomatic:

```
return envoy.Resources{}, fmt.Errorf("failed to inject upstream filters %q (%w): %s", cluster.Name, err, cluster.String())
...
return envoy.Resources{}, fmt.Errorf("failed to validate Cluster after injecting Cilium upstream filters %q (%w): %s", 
```

**Review by mhofstetter — APPROVED** [review]
LGTM - thanks for addressing my input.

Only two use-cases of non-idiomatic error handling left - i assume you're going to fix them.

File: pkg/ciliumenvoyconfig/cec_resource_parser.go (hunk 5 above)
Comment (jrajahalme) [reply pkg/ciliumenvoyconfig/cec_resource_parser.go]: Fixed, thanks for noticing!

**Review by tommyp1ckles — APPROVED** [review]
k8s changes lgtm

**Pull request merged**